taurus/util/tests/test_substitute_links.py

```diff
@@ -0,0 +1,67 @@
+"""Test the substitute_links method, in particular the edge cases that the client doesn't test."""
+import pytest
+from uuid import uuid4
+
+from taurus.util.impl import substitute_links
+from taurus.entity.link_by_uid import LinkByUID
+from taurus.entity.object import MeasurementRun, MaterialRun, ProcessRun, ProcessSpec
+
+
+def test_substitution_without_id():
+    """Test that trying to substitute links if uids haven't been assigned throws an error."""
+    mat = MaterialRun("A material with no id")
+    meas = MeasurementRun("A measurement with no id", material=mat)
+    with pytest.raises(ValueError):
+        substitute_links(meas), "substitute_links should fail if objects don't have uids"
+
+    with pytest.raises(ValueError):
+        substitute_links([meas, mat]), "substitute_links should fail if objects don't have uids"
+
+    with pytest.raises(ValueError):
+        substitute_links(meas.as_dict()), "substitute_links should fail if objects don't have uids"
+
+    # Create a dictionary in which either the key or value is missing a uid
+    meas.add_uid('id', str(uuid4()))
+    with pytest.raises(ValueError):
+        substitute_links({mat: meas}), "substitute_links should fail if objects don't have uids"
+    with pytest.raises(ValueError):
+        substitute_links({meas: mat}), "substitute_links should fail if objects don't have uids"
+
+
+def test_native_id_substitution():
+    """Test that the native id gets serialized, when specified."""
+    native_id = 'id1'
+    other_id = 'id2'
```

Comment:
You might want to add more other_ids so the probability of randomly selecting the `native_id` is lower.

Reply:
Done